### DISCUSSION TITLE
Multi index search

### DISCUSSION THREAD

**Discussion by royduin** [appbaseio/reactivesearch#1721] [Ideas]
It would be nice if the data search component could search in multiple indices so it's possible to search for products, categories and pages at once. For example: https://magento2.algolia.com, try the autocomplete. Any plans for this?

**Comment by siddharthlatest** [comment]
@royduin This is a great idea. We just added support for it in the latest release: 3.19.0, using the `index` prop. (It currently requires use of appbase.io service along with Elasticsearch as it relies on BE query generation. )

**Reply by royduin** [reply]
Any plans for this without the AppBase requirement?

**Reply by siddharthlatest** [reply]
@royduin We're going to separate the commercial dependency of using the server query generation component from ReactiveSearch. The above and some other features, e.g. distinctField that depend on server-side query generation, will not require an appbase.io subscription and work for all users provided they install the server binary as well.

More thoughts on how we will get there:
- We will make the server-side query generation mandatory instead of it being an opt-in today and also open-source this (let's say we call this `reactivesearch-api`)
- This will also allow us to do away with the FE query generation, which isn't secure by default as it allows for arbitrary script injections and opens up the Elasticsearch cluster to expensive DoS attacks. A nice side effect of doing this is also a more compact bundle size.

ReactiveSearch going forward will be the set of FE libraries and the server-side API generation logic. All of these will be open-source, and we will work on making it a single command to add the server-side component to any environment (docker image and a direct binary for Linux environments)

**Reply by royduin** [reply]
But what's wrong with doing multiple queries for each index? The possibility to talk directly to Elasticsearch is a big plus for us and our users; we do not want to force them to use anything besides Elasticsearch.

**Reply by siddharthlatest** [reply]
@royduin The challenge is with the FE query generation piece, it opens up a security vulnerability (a malicious actor can inspect network tabs, generate a different query and either access the data or cause easy DoS attacks) that makes it unsuitable for any production use. 

This has been one of the reasons for our shift in focus to a server side query generation. Currently, this requires using a paid service / proprietary license, we're moving to having a free tier / open-source (Apache 2.0) license for it so that it still remains accessible. (The only change would be installing a Linux package[1] and setting up a reverse proxy to Elasticsearch)

I would love to talk more on how this change can affect you or your users, we definitely want to account for it.

[1] or a docker image

**Reply by royduin** [reply]
With a reverse proxy for SSL and a service like Cloudflare in front of it or with a hosted Elasticsearch service it doesn't have to be an issue. With xpack security it's possible to give users the proper rights and access.

**Reply by siddharthlatest** [reply]
@royduin Here's a scenario that gets missed:

A developer is using ReactiveSearch to power their search app.

Now what the ReactiveSearch powered search app does today is to run a DSL against the Elasticsearch cluster. Any user can inspect the network tab and create a cURL request to run expensive search queries, e.g. anything aggregation heavy or one which fetches a huge number of hits. (Even if the developer has TLS + CF enabled and are using X-Pack, nothing prevents the above from happening)

The key issue is the exposure of the very powerful query DSL directly within FE. As a reference, if a user can run arbitrary SQL queries against a database, it is considered a serious security vulnerability: https://www.w3schools.com/sql/sql_injection.asp.

---

The solution here is two-fold:
- ReactiveSearch should only send declarative inputs, not DSL queries
- For cases where custom queries need to be extended, users can whitelist the queries (and parameterize) and then pass the inputs from ReactiveSearch (i.e. FE client)

This effectively addresses this attack vector. For other vectors, something like CF is good for DDoS and X Pack or similar alternatives can provide a good role-based access control policy.

**Reply by royduin** [reply]
I know but there is currently no free alternative so if we implement multi index searches in Rapidez we're forcing users to a paid solution. Don't get me wrong; I like Appbase but as an open source project we cannot force users.

So for now the only thing we can do is securing Elasticsearch the best we can; SSL, Cloudflare, X-Pack with a read only user, lowering `max_result_window` and keeping an eye on any possible abuse by monitoring.

Any ETA on the server-side component?

**Reply by siddharthlatest** [reply]
@royduin 

We now have the code for server-side query component public over here: https://github.com/appbaseio/reactivesearch-api. It should be released in the next few weeks, we're working on the site and docs related content changes for this.

**Reply by royduin** [reply]
Nice! Watching the repo 😎
